File: Internal/ATTNUserAgentBuilder.m
//
//  ATTNUserAgentBuilder.m
//  attentive-ios-sdk-framework
//
//  Created by Wyatt Davis on 3/14/23.
//

#import <Foundation/Foundation.h>
#import "ATTNUserAgentBuilder.h"
#import "ATTNAppInfo.h"

@implementation ATTNUserAgentBuilder

+ (NSString*)buildUserAgent {
    // We replace the spaces with dashes for the app name because spaces in a User-Agent represent a new "product", so app names that have spaces are harder to parse if we don't replace spaces with dashes
    return [NSString stringWithFormat:@"%@/%@.%@ (%@; %@ %@) %@/%@",
            [self replaceSpacesWithDashes:[ATTNAppInfo getAppName]],
            [ATTNAppInfo getAppVersion],
            [ATTNAppInfo getAppBuild],
            [ATTNAppInfo getDeviceModelName],
            [ATTNAppInfo getDevicePlatform],
            [ATTNAppInfo getDeviceOsVersion],
            [ATTNAppInfo getSdkName],
            [ATTNAppInfo getSdkVersion]];
}

+ (NSString*)replaceSpacesWithDashes:(NSString*)stringToUpdate {
    return [stringToUpdate stringByReplacingOccurrencesOfString:@" " withString:@"-"];
}

@end

File: Sources/Migration/Internal/ATTNUserAgentBuilder.swift
//
//  ATTNUserAgentBuilder.swift
//  attentive-ios-sdk-framework
//
//  Created by Vladimir - Work on 2024-05-31.
//

import Foundation

@objc(ATTNUserAgentBuilder)
open class ATTNUserAgentBuilder: NSObject {

  @objc(buildUserAgent)
  open class func buildUserAgent() -> String {
    buildUserAgent(appInfo: ATTNAppInfo.self)
  }

  @objc
  public static func buildUserAgent(appInfo: ATTNAppInfo.Type) -> String {
    // We replace the spaces with dashes for the app name because spaces in a User-Agent represent a new "product", so app names that have spaces are harder to parse if we don't replace spaces with dashes
    String(
      format: "%@/%@.%@ (%@; %@ %@) %@/%@",
      appInfo.getFormattedAppName(),
      appInfo.getAppVersion(),
      appInfo.getAppBuild(),
      appInfo.getDeviceModelName(),
      appInfo.getDevicePlatform(),
      appInfo.getDeviceOsVersion(),
      appInfo.getSdkName(),
      appInfo.getSdkVersion()
    )
  }
}
